src/pyj/read_book/cfi.pyj

```diff
@@ -694,74 +703,108 @@ def at_point(ox, oy):  # {{{
         Math.sqrt(Math.pow(p1[0]-p2[0], 2), Math.pow(p1[1]-p2[1], 2))
 
-    try:
-        cfi = at(ox, oy)
-        p = point(cfi)
-    except Exception:
-        cfi = None
+    cfi = at(ox, oy)
+    if cfi is None:
+        return None
 
-    if not p:
+    decoded = decode_with_range(cfi)
+    if not decoded:
         return None
 
     if cfi:
-        if p.range is not None:
-            r = p.range
-            rect = r.getClientRects()[0]
-
-            x = (p.a*rect.left + (1-p.a)*rect.right)
-            y = (rect.top + rect.bottom)/2
-            x, y = scroll_viewport.viewport_to_document(x, y, r.startContainer.ownerDocument)
-        else:
-            node = p.node
-            r = node.getBoundingClientRect()
-            # Start of element is right side in RTL, so be sure to get that side in RTL mode
-            x, y = scroll_viewport.viewport_to_document(
-                r.left if scroll_viewport.ltr else r.right, r.top, node.ownerDocument)
-            if jstype(p.x) is 'number' and node.offsetWidth:
-                x += (p.x*node.offsetWidth)/100
-            if jstype(p.y) is 'number' and node.offsetHeight:
-                y += (p.y*node.offsetHeight)/100
-
-        if dist(scroll_viewport.viewport_to_document(ox, oy), v'[x, y]') > 50:
+        cfix, cfiy = decoded_with_range_to_document_position(decoded)
+        if cfix is None or cfiy is None or dist(scroll_viewport.viewport_to_document(ox, oy), v'[cfix, cfiy]') > 50:
             cfi = None
 
     return cfi
 # }}}
 
 def at_current():  # {{{
-    winx, winy = window_scroll_pos()
-    winw, winh = scroll_viewport.width(), scroll_viewport.height()
-    winw = max(winw, 400)
-    winh = max(winh, 600)
-    deltay = Math.floor(winh/50)
-    deltax = Math.floor(winw/25)
-    miny = max(-winy, -winh)
-    maxy = winh
-    minx = max(-winx, -winw)
-    maxx = winw
-
-    def x_loop(cury):
-        for direction in v'[-1, 1]':
-            delta = deltax * direction
-            curx = 0
-            while not ((direction < 0 and curx < minx) or (direction > 0 and curx > maxx)):
-                cfi = at_point(curx, cury)
-                if cfi:
-                    return cfi
-                curx += delta
-
-    for direction in v'[-1, 1]':
-        delta = deltay * direction
-        cury = 0
-        while not( (direction < 0 and cury < miny) or (direction > 0 and cury > maxy) ):
-            cfi = x_loop(cury, -1)
+    wini, winb = scroll_viewport.inline_size(), scroll_viewport.block_size()
+
+    # We subtract one because the the actual position query for CFI elements is relative to the
+    # viewport, and the viewport coordinates actually go from 0 to the size - 1.
+    # If we don't do this, the first line of queries will always fail in those cases where we
+    # start at the right of the viewport because they'll be outside the viewport
+    wini -= 1
+    winb -= 1
+
+    # Don't let the deltas go below 10 or above 30, to prevent perverse cases where
```

Comment:
That should say "5." Worst-case where the entire viewport is scanned, with this set at the minimum of 5, and there is no CFI found was about 1.5 seconds in my tests.